File: backend/node/src/domain/base/interfaces/base-repository.interface.ts
import { BaseEntity } from '../../entities/base.entity';
import { IBasePaginatedResponse } from './base-response.interface';

export interface IBaseReadRepository<Entity extends BaseEntity> {
	read: (entityId: string) => Promise<Entity>;

	readAll: (
		page: number,
		limit: number,
		options: any
	) => Promise<IBasePaginatedResponse<Entity>>;
}

export interface IBaseRepository<Entity extends BaseEntity>
	extends IBaseReadRepository<Entity> {
	create: (data: Entity) => Promise<Entity>;

	update: (entityId: string, data: Partial<Entity>) => Promise<Entity>;

	delete: (entityId: string) => Promise<Entity>;
}

File: backend/go/domain/base/base-repository.interface.go
package base

type IBaseReadRepository[Entity comparable] interface {
	Read(entityId string) (Entity, error)

	ReadAll(page int, limit int, options any) (IBasePaginatedResponse[Entity], error)
}

type IBaseRepository[Entity comparable] interface {
	IBaseReadRepository[Entity]

	Create(data Entity) (Entity, error)
	Update(entityId string, data Entity) (Entity, error)
	Delete(entityId string) (Entity, error)
}
